### ISSUE TITLE
CPU usage +500%

### ISSUE THREAD

**Issue by cadalso** [rundeck/rundeck#715]
Hi,

Our rundeck install is using too much CPU, even when there is no job running. It reaches more than 500% at intervals.

```
top - 22:59:20 up 155 days, 20:00,  1 user,  load average: 3.21, 3.31, 2.60
Tasks:  93 total,   1 running,  92 sleeping,   0 stopped,   0 zombie
%Cpu(s): 81.9 us,  2.2 sy,  0.1 ni, 15.1 id,  0.4 wa,  0.0 hi,  0.2 si,  0.0 st
KiB Mem:  24745688 total, 24463620 used,   282068 free,   284092 buffers
KiB Swap: 31245324 total,  5253424 used, 25991900 free,  9412468 cached

  PID USER      PR  NI  VIRT  RES  SHR S  %CPU %MEM    TIME+  COMMAND                                                           
30382 rundeck   20   0 3009m 1.7g  14m S 523.3  7.1  16:51.93 /usr/bin/java -Djava.security.auth.login.config=/etc/rundeck/jaas 
```

Usually it stays in aprox 290%

```
top - 23:02:03 up 155 days, 20:02,  1 user,  load average: 3.45, 3.51, 2.80
Tasks:  93 total,   1 running,  92 sleeping,   0 stopped,   0 zombie
%Cpu(s): 47.4 us,  2.1 sy,  0.1 ni, 49.7 id,  0.5 wa,  0.0 hi,  0.2 si,  0.0 st
KiB Mem:  24745688 total, 24499556 used,   246132 free,   277052 buffers
KiB Swap: 31245324 total,  5254580 used, 25990744 free,  9082248 cached

  PID USER      PR  NI  VIRT  RES  SHR S  %CPU %MEM    TIME+  COMMAND                                                           
30382 rundeck   20   0 3009m 2.0g  14m S 293.9  8.4  25:02.23 /usr/bin/java -Djava.security.auth.login.config=/etc/rundeck/jaas 
```

RAM configuration is 

```
RDECK_JVM="$RDECK_JVM -Xmx2048m -Xms512m -XX:MaxPermSize=512m -server"
```

Is there any way I can limit rundeck CPU usage? 
Any clues why it is killing the CPU like that?

Thanks

**Comment by bergerx** [comment]
We experienced a similar issue, everything was fine while job is running, but after the job completed Rundeck suddenly started to utilize all the CPUs. Load on the node went up to 15.

When we checked with strace, 4 threads are likely racing for getting the futex and that 4 threads are only reading the logs of recent job (the output of job is ~1G).

```
[doganb@rundeckprod001iad ~]$ sudo strace -c -Ff -p 3430
...
% time     seconds  usecs/call     calls    errors syscall
------ ----------- ----------- --------- --------- ----------------
 70.53    0.225182         235       959       320 futex
 25.98    0.082952        3771        22        18 restart_syscall
  3.45    0.011010         367        30           epoll_wait
  0.01    0.000040           0      1113           clock_gettime
  0.01    0.000038           1        73           gettimeofday
  0.01    0.000036           0       617           read
  0.00    0.000003           0        85           sched_yield
  0.00    0.000000           0        20           mprotect
------ ----------- ----------- --------- --------- ----------------
100.00    0.319261                  2919       338 total
```

There were 4 threads trying to parse the same file:
```
[doganb@rundeckprod001iad ~]$ sudo ls -lrn /proc/3430/fd ?| sort -nk9
...
lr-x------ 1 5006 5006 64 Jul 14 09:05 222 -> /var/lib/rundeck/logs/rundeck/DARCI/job/4be199e4-3ad2-42d5-8231-a4f3a39097b6/logs/97.rdlog
lr-x------ 1 5006 5006 64 Jul 14 08:51 228 -> /var/lib/rundeck/logs/rundeck/DARCI/job/4be199e4-3ad2-42d5-8231-a4f3a39097b6/logs/97.rdlog
lr-x------ 1 5006 5006 64 Jul 14 08:51 230 -> /var/lib/rundeck/logs/rundeck/DARCI/job/4be199e4-3ad2-42d5-8231-a4f3a39097b6/logs/97.rdlog
lr-x------ 1 5006 5006 64 Jul 14 08:51 232 -> /var/lib/rundeck/logs/rundeck/DARCI/job/4be199e4-3ad2-42d5-8231-a4f3a39097b6/logs/97.rdlog
[doganb@rundeckprod001iad ~]$ 
```

**Comment by gschueler** [comment]
@bergerx it would be useful to see a threaddump at that point of high cpu usage

**Comment by snebel29** [comment]
This is interesting that futex stats are high also when there is no activity/high load on the rundeck instance (To let it clear this stats comes from the same instance that @bergerx published)

{code}
% time     seconds  usecs/call     calls    errors syscall
------ ----------- ----------- --------- --------- ----------------
 90.76    1.162078         447      2602      1146 futex
  4.21    0.053946         403       134           epoll_wait
  3.58    0.045805       15268         3           accept
  1.36    0.017406         725        24        18 restart_syscall
  0.03    0.000410           0      4026           clock_gettime
  0.02    0.000287           0      8511           gettimeofday
  0.01    0.000173           0       584           sendto
  0.01    0.000095           2        41           dup2
  0.01    0.000092           0       590           recvfrom
  0.00    0.000052           0      1566           ioctl
  0.00    0.000051           0       208           write
  0.00    0.000037          19         2           getdents
  0.00    0.000003           0        40           connect
  0.00    0.000001           0         4           mprotect
  0.00    0.000001           1         2           writev
  0.00    0.000001           0        14           epoll_ctl
  0.00    0.000000           0        24         1 read
  0.00    0.000000           0         1           open
  0.00    0.000000           0        46           close
  0.00    0.000000           0         8           stat
  0.00    0.000000           0         1           fstat
  0.00    0.000000           0         5           mmap
  0.00    0.000000           0         1           munmap
  0.00    0.000000           0         9           rt_sigreturn
  0.00    0.000000           0        41           socket
  0.00    0.000000           0         3           recvmsg
  0.00    0.000000           0        44         3 shutdown
  0.00    0.000000           0         1           bind
  0.00    0.000000           0        47           getsockname
  0.00    0.000000           0       120           setsockopt
  0.00    0.000000           0         9           fcntl
  0.00    0.000000           0         1           openat
------ ----------- ----------- --------- --------- ----------------
100.00    1.280438                 18712      1168 total
{code}

We will take a thread dump next time this happens :-)

**Comment by gschueler** [comment]
I will close the issue. If you still have this problem, please re-open this issue.

**Comment by drbig** [comment]
Seems we're having an issue that has the same symptoms:
![CPU Usage](https://i.imgur.com/brGn1RP.png)

I've restarted the container at 10:00. What I can tell:
- Rundeck version is `3.1.3-20191204`
- Containerized deployment of image I built (Ubuntu-base, `java-8-openjdk-amd64`; I basically install a .deb and "run with defaults";)
- We're not using "production DB", just the default (`H2`?)
- We're having a single user of Rundeck, i.e. no parallel executions etc.
- This is the second time this happened; it works fine until "it doesn't"; no visible correlation between any user activity, timing, "weight of jobs to run"
- When this happens one can still browse executions etc. but it hangs on running a card (never goes from the params page to the execution page)

I'd gladly take in "things to check/try"; my initial idea was that perhaps lack of prod DB might be it, but reading the threads here I think I'll try to get more data (logs, check threads) when this happens again...
